### PR TITLE
Dynamic exclusion of Trash folder

### PR DESCRIPTION
**Pull request by stephdl**
The Trash folder might be synchronized, depended on the need of the system administrator. The PR intends to synchronize the folder on an administrator input, the default is the Trash sync is disabled

Screenshot (37)
Screenshot (38)
Screenshot (39)
Screenshot (40)

https://github.com/NethServer/dev/issues/6376

### CODE REVIEW

File: api/imapsync/read [hunk 1]
```diff
@@ -128,9 +129,10 @@ if ($cmd eq 'list') {
     my $name = $input->{'name'};
     my $Port = $input->{'Port'};
     my $Security = $input->{'Security'};
-    my $exclude = $input->{'Exclude'};
-    $exclude =~ s/,/|/g;
-    $exclude = '|'.$exclude  if ($exclude ne '');
+    my @exclusion = split(',',$input->{'Exclude'});
+    push (@exclusion, '^Trash') if $input->{'TrashSync'} eq 'disabled';
+    my $exclude = join ('|', @exclusion) || '';
+    $exclude = '|'. $exclude  if ($exclude ne '');
```
Comment (DavidePrincipi) [inline api/imapsync/read:135]: I think the previous implementation was more readable. The patch can be smaller if we just append a line like this after L133:

```perl
$exclude .= '|^Trash' if $input->{'TrashSync'} eq 'disabled';
```

What do you think?

File: api/imapsync/update [hunk 2]
```diff
@@ -75,8 +76,11 @@ if ($cmd eq 'update') {
         $deletefolders = '--delete2folders';
     }
 
-    my $exclude = join ('|',@{$input->{'Exclude'}}) || '';
+    my @exclusion = @{$input->{'Exclude'}};
+    push (@exclusion , '^Trash') if $input->{'TrashSync'} eq 'disabled';
+    my $exclude = join ('|', @exclusion) || '';
     $exclude = '|'.$exclude  if ($exclude ne '');
```
Comment (DavidePrincipi) [inline api/imapsync/update:82]: Same as above, I'd go with a string-append operation. It's just one more line...

File: ui/public/i18n/language.json [hunk 3]
```diff
@@ -495,7 +495,8 @@
         "Synchronization_info":"Synchronization information",
         "synchronization_informations_ok": "Synchronization information has been updated",
         "synchronization_informations_error": "Synchronization information has not been updated",
-        "Exclude": "Exclude folders"
+        "Exclude": "Exclude folders",
+        "TrashSync": "Synchronise the Trash"
```
Comment (DavidePrincipi) [inline ui/public/i18n/language.json:499]: To keep the JSON patch clean, add the new entry before the last one. Also fix the US spelling for `synchroni(s|z)e`

```suggestion
        "TrashSync": "Synchronize the Trash",
        "Exclude": "Exclude folders"
```

**Review by DavidePrincipi — CHANGES_REQUESTED** [review]
Thank you for this new feature! When a patch impacts the UI, could you attach also a screenshot?

Please open also a new issue and link this PR to it.

Code is good however I'd try to keep the patch smaller.

File: api/imapsync/read (hunk 1 above)
Comment (stephdl) [reply api/imapsync/read:135]: need to be tested because 

`||` this will exclude all
`toto|titi|` same

It is hard to have the good behaviour

**Comment by stephdl** [comment]
picture uploaded at the top message

File: api/imapsync/update (hunk 2 above)
Comment (stephdl) [reply api/imapsync/update:80]: done

File: api/imapsync/read (hunk 1 above)
Comment (stephdl) [reply api/imapsync/read]: done good idea

File: ui/public/i18n/language.json [hunk 4]
```diff
@@ -518,6 +519,7 @@
         "lifetime": "A value of 0 means retry only once.",
         "DeleteDestination": "Delete messages in the local server that are not in the remote server. Useful for backup or pre-sync.",
         "imapsync":"Imapsync",
+        "TrashSync":"The exclusion of the Trash folder is done with the exact string '^Trash'. If you have translated it you must add manually the translation to the textarea.",
```
Comment (DavidePrincipi) [inline ui/public/i18n/language.json:522]:
```suggestion
        "TrashSync": "The actual trash can folder name depends on the IMAP client configuration. If the trash can folder name does not start with 'Trash' this option has no effect.",
```

**Review by DavidePrincipi — COMMENTED** [review]
I'd prefer to not tell "what to do if...": If the trash can folder does not start with "Trash" it is synchronized like any other folder. The exclusion field is there for that purpose.0

Instead here I'dexplain the reason why it might not work as expected.

File: api/imapsync/read [hunk 5]
```diff
@@ -145,7 +147,7 @@ if ($cmd eq 'list') {
     my @imapsync =("/usr/bin/imapsync","--host2","localhost","--port2","143","--tls2","--user2",
         $name.'*vmail',"--passfile2","/var/lib/nethserver/imapsync/vmail.pwd","--host1",$hostname,
         "--port1",$Port,$crypt,"--user1",$username,"--passfile1","/var/lib/nethserver/imapsync/".$name.".pwd",
-        "--exclude","^Trash|^Deleted Items|^Public|^Shared".$exclude,
+        "--exclude","^Deleted Items|^Public|^Shared".$exclude,
```
Comment (gsanchietti) [inline api/imapsync/read:150]: I'm not sure, but that `Deleted Items` looks like an alias for `Trash`.

**Comment by gsanchietti** [comment]
There is no a simple implementation to catch all trash folders for different IMAP servers. So I propose this solution:
- enable trash sync by default: removeexclusions for `Trash` and `Deleted Items` folders (the admin should just make sure that original thrash folder is empty if he/she doesn't want to move the garbage to the new server)
- improve the `i` text with something like:  "The actual trash folder name depends on the IMAP client configuration. If the trash  folder name does not start with 'Trash' or 'Deleted Items' this option has no effect.",

**Review by gsanchietti — APPROVED** [review]
It seems good to me, but I would like to wait @DavidePrincipi approval

**Review by DavidePrincipi — APPROVED** [review]
Ok, for me it's a big enhancement! You can merge it as-is :+1:

However I ask if we can push it further and generalize the "default exclusion" concept.

I'm thinking about the "Public" and "Shared" namespaces prefixes: as we are synchronizing with a remote IMAP server we can't make assumptions on its folder naming scheme.

"Synchronize the Trash" could become a switch between the "default exclusion" pattern and a custom one (pre-filled with the default pattern and modifiable).

**Pull request merged**